sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java

```diff
@@ -93,13 +120,155 @@ public void populateDisplayData(DisplayData.Builder builder) {
+            // The writer does not need to be reset, as this DoFn cannot be reused.
+          } catch (Throwable closeException) {
+            // Do not mask the exception that caused the write to fail.
+            e.addSuppressed(closeException);
+          }
+          throw e;
+        }
+      }
+
+      @Override
+      public void finishBundle(Context c) throws Exception {
+        if (writer != null) {
+          WriteT result = writer.close();
+          c.output(result);
+          // Reset state in case of reuse
+          writer = null;
+        }
+      }
+
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        Write.Bound.this.populateDisplayData(builder);
+      }
+    }
+
+    /**
+     * Like {@link WriteBundles}, but where the elements for each shard have been collected into
+     * a single iterable.
+     *
+     * @see WriteBundles
+     */
+    private class WriteShardedBundles<WriteT> extends DoFn<KV<Integer, Iterable<T>>, WriteT> {
+      private final PCollectionView<WriteOperation<T, WriteT>> writeOperationView;
+
+      WriteShardedBundles(PCollectionView<WriteOperation<T, WriteT>> writeOperationView) {
+        this.writeOperationView = writeOperationView;
+      }
+
+      @Override
+      public void processElement(ProcessContext c) throws Exception {
+        // In a sharded write, single input element represents one shard. We can open and close
+        // the writer in each call to processElement.
+        WriteOperation<T, WriteT> writeOperation = c.sideInput(writeOperationView);
+        LOG.info("Opening writer for write operation {}", writeOperation);
+        Writer<T, WriteT> writer = writeOperation.createWriter(c.getPipelineOptions());
+        writer.open(UUID.randomUUID().toString());
+        LOG.debug("Done opening writer {} for operation {}", writer, writeOperationView);
+
+        try {
+          for (T t : c.element().getValue()) {
+            writer.write(t);
+          }
+        } catch (Exception e) {
+          try {
+            writer.close();
+          } catch (Throwable closeException) {
+            // Do not mask the exception that caused the write to fail.
+            e.addSuppressed(closeException);
+          }
+          throw e;
+        }
+
+        // Close the writer; if this throws let the error propagate.
+        WriteT result = writer.close();
+        c.output(result);
+      }
+
+      @Override
+      public void populateDisplayData(DisplayData.Builder builder) {
+        Write.Bound.this.populateDisplayData(builder);
+      }
+    }
+
+    private static class ApplyShardingKey<T> implements SerializableFunction<T, Integer> {
+      private final int numShards;
+      private int shardNumber;
+
+      ApplyShardingKey(int numShards) {
+        this.numShards = numShards;
+        shardNumber = ThreadLocalRandom.current().nextInt(numShards);
+      }
+
+      @Override
+      public Integer apply(T input) {
```

Comment:
Take a look at https://github.com/apache/incubator-beam/blob/master/runners/direct-java/src/main/java/org/apache/beam/runners/direct/ShardControlledWrite.java#L69 for fixing the assignment